src/python/pants/backend/codegen/protobuf/java/rules.py

```diff
@@ -0,0 +1,124 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+
+from pants.backend.codegen.protobuf.protoc import Protoc
+from pants.backend.codegen.protobuf.target_types import ProtobufSourceField
+from pants.backend.python.target_types import PythonSourceField
+from pants.backend.python.util_rules import pex
+from pants.core.util_rules.external_tool import DownloadedExternalTool, ExternalToolRequest
+from pants.core.util_rules.source_files import SourceFilesRequest
+from pants.core.util_rules.stripped_source_files import StrippedSourceFiles
+from pants.engine.fs import (
+    AddPrefix,
+    CreateDigest,
+    Digest,
+    Directory,
+    MergeDigests,
+    RemovePrefix,
+    Snapshot,
+)
+from pants.engine.platform import Platform
+from pants.engine.process import Process, ProcessResult
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import (
+    GeneratedSources,
+    GenerateSourcesRequest,
+    TransitiveTargets,
+    TransitiveTargetsRequest,
+)
+from pants.engine.unions import UnionRule
+from pants.source.source_root import SourceRoot, SourceRootRequest
+from pants.util.logging import LogLevel
+
+
+class GenerateJavaFromProtobufRequest(GenerateSourcesRequest):
+    input = ProtobufSourceField
+    output = PythonSourceField
+
+
+@rule(desc="Generate Java from Protobuf", level=LogLevel.DEBUG)
+async def generate_java_from_protobuf(
+    request: GenerateJavaFromProtobufRequest,
+    protoc: Protoc,
+) -> GeneratedSources:
+    download_protoc_request = Get(
+        DownloadedExternalTool, ExternalToolRequest, protoc.get_request(Platform.current)
+    )
+
+    output_dir = "_generated_files"
+    create_output_dir_request = Get(Digest, CreateDigest([Directory(output_dir)]))
+
+    # Protoc needs all transitive dependencies on `protobuf_libraries` to work properly. It won't
```

Comment:
I imagine this was copy pasta, but, stale. Should be `protobuf_source`.

Reply:
https://github.com/pantsbuild/pants/pull/14006